cpp/src/join/mixed_join_kernels_semi.cu

```diff
@@ -52,24 +56,24 @@ CUDF_KERNEL void __launch_bounds__(block_size)
   cudf::ast::detail::IntermediateDataType<has_nulls>* intermediate_storage =
     reinterpret_cast<cudf::ast::detail::IntermediateDataType<has_nulls>*>(raw_intermediate_storage);
   auto thread_intermediate_storage =
-    &intermediate_storage[threadIdx.x * device_expression_data.num_intermediates];
-
-  cudf::size_type const left_num_rows  = left_table.num_rows();
-  cudf::size_type const right_num_rows = right_table.num_rows();
-  auto const outer_num_rows            = left_num_rows;
+    &intermediate_storage[tile.meta_group_rank() * device_expression_data.num_intermediates];
 
-  cudf::size_type outer_row_index = threadIdx.x + blockIdx.x * block_size;
+  cudf::size_type const outer_num_rows  = left_table.num_rows();
+  cudf::size_type const outer_row_index = (threadIdx.x + blockIdx.x * block_size) / cg_size;
```

Comment:
```suggestion
  cudf::size_type const outer_num_rows  = left_table.num_rows();
  auto const outer_row_index = cudf::detail::grid_1d::global_thread_id<block_size>() / cg_size;
```
https://github.com/rapidsai/cudf/blob/3dbc33a5cb1cf7052cd67f5654b34594403fbfef/cpp/include/cudf/detail/utilities/cuda.cuh#L91-L143
Ideally, we need to always use `grid_1d` to calculate all global thread indices and strides.